Module command result callback addition

martinrvisser opened the pull request:
## Add Command Result Event Notifications for Modules

### Summary

1. Adds new server events `ValkeyModuleEvent_CommandResultSuccess` and `ValkeyModuleEvent_CommandResultFailure` for that can notify subscribed modules after command execution. This enables modules to implement audit logging, error monitoring, performance tracking, and observability without modifying core server code.
2. Adds new server event `ValkeyModuleEvent_CommandResultACLDenied` for commands rejected by ACL. Together with PR #2237 this covers auditing of authentication and authorisation. 

### Motivation

There is currently no module API to observe command outcomes after execution or to capture ACL denied commands. Modules that need audit logging or error monitoring have no mechanism to be notified when commands succeed or fail, what arguments were used, how long they took, or how many keys were modified. This feature fills that gap using the existing `ValkeyModule_SubscribeToServerEvent()` infrastructure.

### API

#### Events

| Event | Description |
|---|---|
| `ValkeyModuleEvent_CommandResultSuccess` | Fired after a command completes successfully |
| `ValkeyModuleEvent_CommandResultFailure` | Fired after a command returns an error |
| `ValkeyModuleEvent_CommandACLDenied` | Fired after a command is rejected by ACL |

These are separate events (not sub-events), so modules can for example only subscribe to failures without incurring any callback overhead for successful commands.

#### Event Data: `ValkeyModuleCommandResultInfo`

The `data` pointer passed to the callback can be cast to `ValkeyModuleCommandResultInfo`:

```c
typedef struct ValkeyModuleCommandResultInfo {
    uint64_t version;           /* Version of this structure for ABI compat. */
    const char *command_name;   /* Full command name (e.g., "SET", "CLIENT|LIST"). */
    long long duration_us;      /* Execution duration in microseconds. */
    long long dirty;            /* Number of keys modified. */
    uint64_t client_id;         /* Client ID that executed the command. */
    int is_module_client;       /* 1 if command was from RM_Call, 0 otherwise. */
    int argc;                   /* Number of command arguments. */
    ValkeyModuleString **argv;  /* Command arguments array (zero-copy, read-only). */
    int acl_deny_reason;        /* ACL_DENIED_CMD/KEY/CHANNEL/AUTH; 0 for non-ACL events */
    const char *acl_object;     /* Denied resource name (key/channel); NULL for CMD/AUTH */
} ValkeyModuleCommandResultInfoV1;
```

The struct is versioned (`VALKEYMODULE_COMMANDRESULTINFO_VERSION`) for forward-compatible API evolution.

### Usage Example

```c
/* Callback receives events for whichever event(s) you subscribed to */
void OnCommandResult(ValkeyModuleCtx *ctx, ValkeyModuleEvent eid,
                     uint64_t subevent, void *data) {
    VALKEYMODULE_NOT_USED(ctx);
    VALKEYMODULE_NOT_USED(subevent);

    ValkeyModuleCommandResultInfo *info = (ValkeyModuleCommandResultInfo *)data;
    if (info->version != VALKEYMODULE_COMMANDRESULTINFO_VERSION) return;

    int failed = (eid.id == VALKEYMODULE_EVENT_COMMAND_RESULT_FAILURE);

    /* Access fields directly */
    printf("command=%s status=%s duration=%lldus dirty=%lld client=%llu\n",
           info->command_name,
           failed ? "FAIL" : "OK",
           info->duration_us,
           info->dirty,
           info->client_id);

    /* Access argv (read-only, zero-copy) */
    for (int i = 0; i < info->argc; i++) {
        size_t len;
        const char *arg = ValkeyModule_StringPtrLen(info->argv[i], &len);
        printf("  argv[%d] = %.*s\n", i, (int)len, arg);
    }
}

/* Subscribe in ValkeyModule_OnLoad or at runtime */

/* Option A: command failures only (recommended for audit logging) */
ValkeyModule_SubscribeToServerEvent(ctx,
    ValkeyModuleEvent_CommandResultFailure, OnCommandResult);

/* Option B: command successes only */
ValkeyModule_SubscribeToServerEvent(ctx,
    ValkeyModuleEvent_CommandResultSuccess, OnCommandResult);

/* Option C: both command outcomes*/
ValkeyModule_SubscribeToServerEvent(ctx,
    ValkeyModuleEvent_CommandResultSuccess, OnCommandResult);
ValkeyModule_SubscribeToServerEvent(ctx,
    ValkeyModuleEvent_CommandResultFailure, OnCommandResult);

/* Subscribe to ACL Denied */
ValkeyModule_SubscribeToServerEvent(ctx,
        ValkeyModuleEvent_CommandResultACLDenied, onCommandResult);

/* Unsubscribe pass NULL callback */
ValkeyModule_SubscribeToServerEvent(ctx,
    ValkeyModuleEvent_CommandResultFailure, NULL);
```

### Design Decisions

- **Separate events instead of sub-events**: Modules subscribing only to failures have zero overhead for successful commands (~2ns listener-list check vs ~30ns callback invocation per command). This is critical since success events fire on the hot path of every command.
- **Stack-allocated info struct**: The `ValkeyModuleCommandResultInfoV1` is built on the stack ΓÇö no heap allocation per event.
- **Zero-copy argv**: Arguments are passed directly from the client's argv array. Any integer-encoded arguments (from `tryObjectEncoding()` during command execution) are decoded to string-encoded objects before being passed to the callback, ensuring compatibility with `ValkeyModule_StringPtrLen()`.
- **Early exit**: If no modules are subscribed to any server events, the event firing function returns immediately before building the info struct.
- **Uses existing server event infrastructure**: Follows the `ValkeyModule_SubscribeToServerEvent()` pattern used by all other server events, rather than introducing a new callback mechanism.

### Files Changed

| File | Change |
|---|---|
| `src/valkeymodule.h` | Event IDs, event constants, `ValkeyModuleCommandResultInfoV1` struct |
| `src/module.c` | `moduleFireCommandResultEvent()`, event documentation, event version entries |
| `src/module.h` | Function declaration |
| `src/server.c` | Call `moduleFireCommandResultEvent()` from `call()` after command execution |
| `src/server.c` | Call to `moduleFireCommandACLDeniedEvent` in `processCommand` after ACL rejection |
| `tests/modules/commandresult.c` | Test module exercising the full API |
| `tests/unit/moduleapi/commandresult.tcl` | Integration tests |

martinrvisser committed:
Commandresult module callback. Signed-off-by: martinrvisser <mvisser@hotmail.com>

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser commented:
> > `valkey-benchmark` is aimed at benchmarking Valkey commands where as this is specific to module implementations. There are no other module API tests in `valkey-benchmark`?
> 
> Although valkey-benchmark does not have any module specific tests, I still feel that it's the right place to put any kind of benchmark. We need to make a few changes though because we will have to build the module along with valkey-benchmark.
> 
> Since moving benchmark code to valkey-benchmark requires a significant amount of work, and is orthogonal to the main focus of this PR, which is the proposal of a new module API, I propose to remove the benchmarking code from this PR, and just keep unit tests that test the new module API. And then open a new PR with just the benchmarking part.

@rjd15372 Removed the benchmark code

rjd15372 commented on src/module.c:
```diff
@@ -11346,6 +11246,164 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+}
+
+/* Get the number of dirty (modified) keys from a command result context. */
+long long VM_CommandResultDirty(ValkeyModuleCommandResultCtx *rctx) {
```
I think it's better to rename the this function to `VM_CommandResultNumDirtyKeys` to become self explanatory

rjd15372 commented on src/module.c:
```diff
@@ -11346,6 +11246,164 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+        zfree(result);
+    }
+    return count;
+}
```
Can't we implement this in a different way without requiring the `result_callbacks` list?

Something like:

```suggestion
int moduleUnregisterCommandResultCallbacks(ValkeyModule *module) {
    listIter li;
    listNode *ln;
    int count = 0;

    listRewind(moduleCommandResultCallbacks, &li);
    while ((ln = listNext(&li))) {
        ValkeyModuleCommandResult *result = ln->value;
        if (result->module == module) {
            listDelNode(moduleCommandResultCallbacks, ln);
            count++;
        }
        zfree(result);
    }
    return count;
}
```

rjd15372 commented on src/module.c:
```diff
@@ -14637,6 +14661,13 @@ void moduleRegisterCoreAPI(void) {
+    REGISTER_API(CommandResultCommandName);
+    REGISTER_API(CommandResultDuration);
+    REGISTER_API(CommandResultDirty);
+    REGISTER_API(CommandResultClientId);
```
I think we should rename these function to:

```suggestion
    REGISTER_API(CommandResultGetStatus);
    REGISTER_API(CommandResultGetCommandName);
    REGISTER_API(CommandResultGetDuration);
    REGISTER_API(CommandResultGetNumDirtyKeys);
    REGISTER_API(CommandResultGetClientId);
```

rjd15372 commented on src/module.h:
```diff
@@ -105,6 +105,7 @@ typedef struct ValkeyModule {
     list *usedby;                         /* List of modules using APIs from this one. */
     list *using;                          /* List of modules we use some APIs of. */
     list *filters;                        /* List of filters the module has registered. */
+    list *result_callbacks;               /* List of command result callbacks the module has registered. */
```
If you follow my suggestion in `moduleUnregisterCommandResultCallbacks` function, we can remove this field

rjd15372 commented on tests/modules/commandresult.c:
```diff
@@ -0,0 +1,347 @@
+    ValkeyModule_FreeCallReply(reply);
+    ValkeyModule_FreeString(ctx, value);
+    return ValkeyModule_ReplyWithSimpleString(ctx, "OK");
+}
```
I don't think we need this command. The TCL script can just call `r SET ss 3`.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        r cmdresult.register all
+
+        # This command modifies a key
+        r cmdresult.dirty mykey
```
Instead of using this custom module command, just issue a few valkey commands that change keys

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert_equal [llength $log] 2
+
+        r cmdresult.unregister
+    }
```
Not sure if it's worth to be testing the implementation of the test module. The test module is not supposed to be used outside of tests.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert_match {*already registered*} $err
+
+        r cmdresult.unregister
+    }
```
I don't understand this test. The restriction of allowing a single callback to be registered is from the test module implementation and not from the valkey module API.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+
+        catch {r cmdresult.register invalid_flag} err
+        assert_match {*invalid flags*} $err
+    }
```
This is the same situation as before. The module API implementation does not verify the flags that are passed in `ValkeyModule_RegisterCommandResult`. This test is validating just the test module implementation.

Shouldn't the invalid flag error be raised by the module API implementation?

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[dict get $entry client_id] > 0}
+
+        r cmdresult.unregister
+    }
```
We already have the test `Module commandresult - Client ID is captured` above that is testing the same thing. This should be removed.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[dict get $stats total_callbacks] >= 5}
+
+        r cmdresult.unregister
+    }
```
In test `Module commandresult - Multiple callbacks from different operations` we already validate the same thing. I don't think we need this test case.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[dict get $entry dirty] >= 1}
+
+        r cmdresult.unregister
+    }
```
We already test the dirty keys tracking in another test.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[llength $log] <= 100}
+
+        r cmdresult.unregister
+    }
```
I don't think we need to test the log overflow

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[llength $log] <= 6} ;# 5 pings + stats command
+
+        r cmdresult.unregister
+    }
```
Same here.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert_equal [dict get $stats failure_count] 1
+
+        r cmdresult.unregister
+    }
```
Thes two test cases are redundant in the face of all the above test cases.

rjd15372 commented on tests/unit/moduleapi/commandresult.tcl:
```diff
@@ -0,0 +1,561 @@
+        assert {[dict get $stats total_callbacks] >= 1}
+
+        r cmdresult.unregister
+    }
```
Same here

rjd15372 commented on tests/unit/moduleapi/misc.tcl:
```diff
@@ -124,6 +124,17 @@ start_server {overrides {save {900 1}} tags {"modules"}} {
         assert { [dict get $info flags] == "${ssl_flag}:::::" }
     }
 
+    test {test module clientinfo api - primary/replica/monitor flags} {
+        # Normal client should not have primary, replica, or monitor flags
+        set info [r test.clientinfo]
+        set flags [parse_client_flags [dict get $info flags]]
+        assert { "primary" ni $flags }
+        assert { "replica" ni $flags }
+        assert { "monitor" ni $flags }
+        assert { "module" ni $flags }
+        assert { "fake" ni $flags }
+    }
+
```
Why are you adding this unrelated test case?

martinrvisser replied:
Removed

rjd15372 replied:
It's still here

rjd15372 reviewed (changes requested):
Besides my inline comments, I also have a high level question. Should we add the restriction that a module can only register a single command result callback?

I'm not seeing a use case where a module needs to register more than one callback. And if that is the case, then we don't need to store the callbacks on a list, and we can store the module's callback in the module structure.

martinrvisser committed:
Update src/module.c

Co-authored-by: Ricardo Dias <rjd15372@gmail.com>
Signed-off-by: martinrvisser <martinrvisser@users.noreply.github.com>

martinrvisser committed:
Update src/module.c

Co-authored-by: Ricardo Dias <rjd15372@gmail.com>
Signed-off-by: martinrvisser <martinrvisser@users.noreply.github.com>

martinrvisser committed:
Addressed PR comments

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
module.c fix

Signed-off-by: martinrvisser <mvisser@hotmail.com>

rjd15372 commented on src/module.c:
```diff
@@ -11346,6 +11243,164 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+                      VALKEYMODULE_CMDRESULT_NOSELF;
+    if (flags & ~valid_flags) {
+        return NULL;  /* Invalid flags */
+    }
```
In both error paths, we should at least log in the server log the reason why we've fail to register a callback, otherwise it will be hard for the module developer to debug the problem.

zuiderkwast commented on src/valkeymodule.h:
```diff
@@ -2019,6 +2038,18 @@ VALKEYMODULE_API int (*ValkeyModule_CommandFilterArgDelete)(ValkeyModuleCommandF
+VALKEYMODULE_API long long (*ValkeyModule_CommandResultGetDuration)(ValkeyModuleCommandResultCtx *rctx) VALKEYMODULE_ATTR;
+VALKEYMODULE_API long long (*ValkeyModule_CommandResultGetDirty)(ValkeyModuleCommandResultCtx *rctx) VALKEYMODULE_ATTR;
+VALKEYMODULE_API unsigned long long (*ValkeyModule_CommandResultGetClientId)(ValkeyModuleCommandResultCtx *rctx)
+    VALKEYMODULE_ATTR;
```
This looks like a useful feature. The APIs look good in general, but I have a few suggestions:

* `ValkeyModule_RegisterCommandResult` and `ValkeyModule_UnegisterCommandResult`: These names suggest that you register a single result, but what you actually register a callback, or we could call it a result handler. Let's add a suffix like Func, Callback, Handler or similar to the name, e.g. `ValkeyModule_RegisterCommandResultHandler`.

* The type `ValkeyModuleCommandResult` is not a single result of a command. I suggest renaming it to `ValkeyModuleCommandResultHandler` or similar.
* The type `ValkeyModuleCommandResultCtx` represents a single result? Then it could be called `ValkeyModuleCommandResult` instead.

* `ValkeyModule_CommandResultGetDirty`: Let's rename it to `VM_CommandResultNumDirtyKeys` as Ricardo suggested.

* Would it be possible to get the actual reply out of a command's result, for example in the form of a `ValkeyModuleCallReply`? I don't have a use case for it but I image it'd be powerful, possibly a future addition? (e.g. `VM_CommandResultGetReply`)

* Does the result also have access to the original command arguments? If we expose that too, it's not only the result but more like a complete command monitoring. Maybe it should be called command monitor instead?

Please update the PR top comment with the correct API matching what's implemented. Currently, some of them appear to return non-module-API types `client *` and `struct serverCommand *`, which isn't what's implemented:

```C
client *VM_CommandResultGetClient(ValkeyModuleCommandResultCtx *ctx);
struct serverCommand *VM_CommandResultGetCommand(ValkeyModuleCommandResultCtx *ctx);
```

martinrvisser replied:
@zuiderkwast great suggestion! If we add both the command arguments and the reply this would provide a much better place for auditing (as well as the monitoring use case you mention) @rjd15372 thoughts on adding these?

rjd15372 replied:
Having both the cmd args and the reply would be great, but I fear that having the reply might introduce some perf overhead because it will require to copy the resp reply string from the client out buffer and then parse it to create the ValkeyModuleCalReply object.

But without doing proper testing I don't know how much it will cost.

zuiderkwast replied:
If we copy it on demand when the user accesses it using a GetReply call, then it's maybe very cheap if it's not used. I guess we need to keep track of where it starts and ends in the reply buffer and reply list though.

rjd15372 replied:
That's a good idea. In that case I think it's worthwhile to develop these suggestions you've made.

martinrvisser committed:
addition of module API functions to get command arguments and reply

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
resolve int handlin in VM_StringPtrLen after unstable merge

Signed-off-by: martinrvisser <mvisser@hotmail.com>

murphyjacob4 commented on src/module.c:
```diff
@@ -11368,6 +11402,287 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+ * Returns a ValkeyModuleCommandResult pointer that can be used with
+ * ValkeyModule_UnregisterCommandResult().
+ */
+ValkeyModuleCommandResult *VM_RegisterCommandResult(ValkeyModuleCtx *ctx,
```
Any reason not to use `VM_SubscribeToServerEvent`? This could just be a new event type, right?

martinrvisser replied:
@murphyjacob4 SubscribeToServerEvent was not used because :
1. server events are designed to be used for infrequent system events rather than commands. For this reason CommandFilter isn't part of ServerEvent either.
2. for the audit requirement, the FAILURES_ONLY flag is crucial. Currently ServerEvents don't support flags.
3. server events go through each listener for each command and callback wheras with commandresult we skip early depending on the flags
4. serverEvents callbacks receive a ValkeyModuleEvent + void *data so lacks typed accessors like GetArgv

murphyjacob4 replied:
> server events are designed to be used for infrequent system events rather than commands. For this reason CommandFilter isn't part of ServerEvent either.

Yeah I think CommandFilter does a lot more than just notify. It allows modules to rewrite the args too. So not exactly a great parallel. Regarding the design for infrequent system events - this is true that although you can subscribe to broad categories, more refined subscriptions aren't possible. The existing server events are a trickle and command callbacks will be a firehose

> for the audit requirement, the FAILURES_ONLY flag is crucial. Currently ServerEvents don't support flags.

That makes sense - you need the option to select just a subset of commands.

> server events go through each listener for each command and callback wheras with commandresult we skip early depending on the flags

Yeah I guess this is a limitation of the current server events API.

> serverEvents callbacks receive a ValkeyModuleEvent + void *data so lacks typed accessors like GetArgv

Although it is void* you cast it to a struct that is defined in valkeymodule.h and the struct is versioned so that we don't need to introduce new APIs for each new field.

> SubscribeToServerEvent was not used

So it sounds like the primary reason is:

1. Concerns around efficiency
2. Ability to scope notifications down (failures only, or all)

I guess 2 could be modeled as two different subscriptions (one for failure, one for non-failure).

I am mostly hesitant to add a bunch of dedicated APIs since it feels like we should only really maintain one "server callback" API system. I'm not convinced we necessarily need a dedicated one for commands. If the existing "server callback" API needs work (e.g. to support notification masks), I would be more inclined to add a new way to register callbacks with such masks rather than adding a new API for each notification type.

I think if we had a high-performance server callback API with support for sub-event masks it sounds like this is exactly what you want. Maybe we should just build that, so that we don't rebuild this infrastructure for the next high-volume event requirement? What do you think?

martinrvisser replied:
@murphyjacob4 @zuiderkwast maintaining one callback system makes sense even if currently serverEvents is mainly used for infrequent events. 

So we would add VALKEYMODULE_EVENT_COMMAND_RESULT to ValkeyModule_SubscribeToServerEvent with subevents:
```
/* Sub-events */
#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_SUCCESS 0
#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_FAILURE 1
```

Since the main requirement for this PR is for auditing, I propose we minimise the API to : 
- VM_CommandResultGetCommandName(data) : command name
- VM_CommandResultGetDuration(data) : execution duration in microseconds
- VM_CommandResultGetDirty(data) : number of keys modified
- VM_CommandResultGetClientId(data) : client ID
- VM_CommandResultIsModuleClient(data) : if command was from RM_Call
- VM_CommandResultGetArgv(data, &argv, &argc) : original command arguments (zero-copy)

rjd15372 replied:
Since we're following the server events approach, why don't we just expose the structure with the event fields, like we do for other events, instead of having the above accessory functions?

rjd15372 replied:
I also think that improving the efficiency/performance of the server events system can be done in a different PR.

madolson commented on src/server.c:
```diff
@@ -3861,14 +3861,19 @@ void call(client *c, int flags) {
     /* Update failed command calls if required. */
 
-    if (!incrCommandStatsOnError(real_cmd, ERROR_COMMAND_FAILED) && c->deferred_reply_errors) {
+    int command_failed = incrCommandStatsOnError(real_cmd, ERROR_COMMAND_FAILED);
```
There are some cases where commands finish but don't go through this call back. For example, ACL errors, Clustering errors, module command unblocks. We should decide if that should flow through.

martinrvisser replied:
@madolson @rjd15372 added the ACL Denied event

murphyjacob4 replied:
Other than ACL, there are quite a few scenarios, e.g. if you look at processCommand you can see all the rejected scenarios where we call reject command.

I think it is fine if not all of these are captured for now. But I guess ACL is just one class of "rejected" command. Do you think we should extend the ACL even to be a "rejected" event? I guess for now we can just explicitly document that these other rejections don't have module events yet.

martinrvisser replied:
@murphyjacob4 agreed there are quite a few "rejected" classes and from an audit perspective these would be of interest. Include these in this PR (since it was aimed at ACL) or a separate?

murphyjacob4 replied:
I think the question is two-fold:

1. Is the API we are exposing extensible enough to add the other rejected classes? E.g. instead of:

```
ValkeyModuleEvent_CommandResultSuccess | Fired after a command completes successfully
ValkeyModuleEvent_CommandResultFailure | Fired after a command returns an error
ValkeyModuleEvent_CommandACLDenied | Fired after a command is rejected by ACL
```

Should it be:

```
ValkeyModuleEvent_CommandResultSuccess | Fired after a command completes successfully
ValkeyModuleEvent_CommandResultFailure | Fired after a command returns an error
ValkeyModuleEvent_CommandResultRejected | Fired after a command is rejected
```

And we can use the subevent to indicate what type of rejected notification?

2. Are we structuring the code such that future module callbacks won't be missed? There are a lot of exit paths in the function for rejecting commands. AFAIU they all trigger `rejectCommand`. Perhaps we should make the module event firing a side effect of that call?

If we go with the suggestion in 1, I am okay with just firing a generic "rejected" event for non-ACL reject scenarios and we can later add classifications for them.

martinrvisser replied:
Re: 1 the second option is a logical and clean API with the subevents denoting the type
```
ValkeyModuleEvent_CommandResultSuccess | Fired after a command completes successfully
ValkeyModuleEvent_CommandResultFailure | Fired after a command returns an error
ValkeyModuleEvent_CommandResultRejected | Fired after a command is rejected
```
Re: 2 rejectCommand only gets the client struct and error msg string so lacks the subevent. That begs the question, should errors have a type i.e. authRequired and so on?  Since that's perhaps too big a change to consider here, we can go with a wrapper function in processCommand which does the rejectCommand as well as moduleFireCommandRejectedEvent with the subevent.

murphyjacob4 replied:
For 2, it is just a suggestion, I'm also fine if we just call the function in each return path for now. Alternatively we could parameterize rejectCommand with the subevent, or use a `goto` or something to `goto rejected` in the processCommand function. I don't have strong opinions, will leave it up to you. If wrapping isn't too much work, that is also viable, although I don't want to force a large refactor if it isn't needed.

martinrvisser replied:
@murphyjacob4 implemented the general ValkeyModuleEvent_CommandResultRejected with subevents

murphyjacob4 replied:
I like it! We should get one more round of approval from the TSC with the new API

martinrvisser committed:
implementation as serverEvent

Signed-off-by: martinrvisser <mvisser@hotmail.com>

rjd15372 commented on src/module.c:
```diff
@@ -2997,6 +3002,14 @@ const char *VM_StringPtrLen(const ValkeyModuleString *str, size_t *len) {
+        size_t l = ll2string(vm_string_int_buffer, sizeof(vm_string_int_buffer), (long)objectGetVal(str));
+        if (len) *len = l;
+        return vm_string_int_buffer;
+    }
```
Why are you doing this? Why would a `ValkeyModuleString` point to an integer encoded robj?

martinrvisser replied:
@rjd15372 It impacts performance for commands that have integer arguments to convert at source compared to the pre-allocated memory lazy approach. However it's probably worth sticking with the safer existing approach rather than implementing this. WDYT?

rjd15372 replied:
I think we have a bigger problem here. ValkeyModuleString type assumes that the `robj` is pointing to either a `OBJ_ENCODING_EMBSTR` string or a `OBJ_ENCODING_RAW` string. But in this case, the processing of a command, converts string parameters that contain numbers into `OBJ_ENCODING_INT`, which is not supported by `ValkeyModuleString`.

We should have some way to get the integer number from a ValkeyModuleString without converting it back to a string.

@zuiderkwast what are your thoughts on this?

zuiderkwast replied:
So far, ValkeyModuleString was always RAW or EMBSTR, never INT encoded. Is this new INT encoded string coming from the event in this PR? Which one? I don't think a command `argv` is ever INT encoded, but I may be wrong.

If someone wants to take the char pointer of multiple such strings as in the code below, there is a problem because there's only one global buffer.

```C
    printf("%s %s",
           ValkeyModule_StringPtrLen(a, NULL),
           ValkeyModule_StringPtrLen(b, NULL));
```

There can also be other strange cases that we're not aware of. I think it's better to avoid introducing the INT encoding, if possible.

martinrvisser replied:
Agree, we leave VM_StringPtrLen as is to only support OBJ_ENCODING_EMBSTR and OBJ_ENCODING_RAW and move the INT decoding to moduleFireCommandResultEvent.

rjd15372 commented on src/module.c:
```diff
@@ -11368,6 +11381,45 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+
+    /* Determine sub-event based on success/failure */
+    int subevent = command_failed ? VALKEYMODULE_SUBEVENT_COMMAND_RESULT_FAILURE
+                                  : VALKEYMODULE_SUBEVENT_COMMAND_RESULT_SUCCESS;
```
By using subevents to distinguish failures from successes, it's not possible to avoid firing events on successes, which will incur in a big perf overhead.

Should we consider instead to create two different top level events to distinguish between success and failed results?

martinrvisser replied:
@rjd15372 agreed we are better of with two different top level events since it allows the option to subscribe to failures only and avoiding the overhead as you suggested.

martinrvisser committed:
moving from subevent to toplevel event

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
removing INT handling from VM_StringPtrLen, added to moduleFireCommandResultEvent

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
fix build failure

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
format fix

Signed-off-by: martinrvisser <mvisser@hotmail.com>

rjd15372 commented on src/module.c:
```diff
@@ -11368,6 +11369,72 @@ unsigned long long VM_CommandFilterGetClientId(ValkeyModuleCommandFilterCtx *fct
+                                  long long dirty) {
+    /* Fast path: skip if no modules are subscribed to any events.
+     * This avoids building the info struct when there are no listeners. */
+    if (listLength(ValkeyModule_EventListeners) == 0) return;
```
I think we need to improve this fast path check. Even if there are listeners registered it does not mean that there are listeners registered for the COMMAND_RESULT_* events.

If some module registers a listener to listen for some other event, we will go through the slow-path, which might even do allocations.

I think we should iterate the listeners list and check if there is at least one listener listening for COMMAND_RESULT events.

martinrvisser replied:
@rjd15372 implemented counters for success and failure registrations to avoid the slow path

rjd15372 reviewed (commented):
Apart for the fast-path comment, everything else looks good to me.

If after addressing my comment, all tests pass, then I'll approve this PR.

martinrvisser committed:
optimising the fastpath using commandlistenercounts for failure and success

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
clang formatting fix

Signed-off-by: martinrvisser <mvisser@hotmail.com>

rjd15372 reviewed (approved):
LGTM

martinrvisser committed:
Add ValkeyModuleEvent_CommandResultACLDenied for NOPERM denials

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
Clang format fix

Signed-off-by: martinrvisser <mvisser@hotmail.com>

murphyjacob4 commented on src/module.c:
```diff
@@ -12226,6 +12410,7 @@ int VM_IsSubEventSupported(ValkeyModuleEvent event, int64_t subevent) {
     case VALKEYMODULE_EVENT_EVENTLOOP: return subevent < _VALKEYMODULE_SUBEVENT_EVENTLOOP_NEXT;
     case VALKEYMODULE_EVENT_CONFIG: return subevent < _VALKEYMODULE_SUBEVENT_CONFIG_NEXT;
     case VALKEYMODULE_EVENT_KEY: return subevent < _VALKEYMODULE_SUBEVENT_KEY_NEXT;
+    /* VALKEYMODULE_EVENT_COMMAND_RESULT_* events have no sub-events */
```
nit: other events that don't need subevents don't add any comments here. I think it is pretty clear to me that the default is no subevent. Should we follow suit?

murphyjacob4 commented on src/server.c:
Do we need to cover this branch for ACL denied as well?

martinrvisser replied:
@murphyjacob4 thanks for this! added

murphyjacob4 reviewed (commented):
Overall makes sense, just want to hone in on the callback scenarios

martinrvisser committed:
Adjusted code comments and added addition firing of ACL denied

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
moving from just ACLdenied to general rejections with subevents

Signed-off-by: martinrvisser <mvisser@hotmail.com>

murphyjacob4 commented on src/module.c:
```diff
@@ -12162,6 +12292,64 @@ static uint64_t moduleEventVersions[] = {
+ *
+ *     Called when a command is rejected before execution. The subevent argument
+ *     identifies the rejection reason (VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_*).
+ *     All 19 rejection paths in processCommand fire this event, including:
```
```suggestion
 *     Rejection paths in processCommand fire this event, including:
```

So the comment doesn't become stale

martinrvisser replied:
Done

murphyjacob4 commented on src/module.c:
```diff
@@ -12162,6 +12292,64 @@ static uint64_t moduleEventVersions[] = {
+ *       - UNKNOWN_CMD / WRONG_ARITY: invalid command syntax
+ *       - OOM, LOADING, BUSY, STALE, DISK_ERROR, NO_REPLICAS, RO_REPLICA, etc.
+ *
+ *     For NOAUTH, see also ValkeyModuleEvent_AuthenticationAttempt (PR #2237)
```
nit

```suggestion
 *     For NOAUTH, see also ValkeyModuleEvent_AuthenticationAttempt
```

The commit history can be found via the git blame

murphyjacob4 commented on src/module.c:
```diff
@@ -12162,6 +12292,64 @@ static uint64_t moduleEventVersions[] = {
+ *       - ACL_KEY / ACL_CHANNEL: the denied key or channel name
+ *       - UNKNOWN_CMD: the attempted command string (command_name is NULL in this case)
+ *       - CLUSTER / REDIRECT: the redirect target address as "host:port"
+ *       - All other subevents: NULL
```
nit: I think we can just document this in the struct itself, to avoid two places to update comments on modification

murphyjacob4 commented on src/valkeymodule.h:
```diff
@@ -839,6 +866,23 @@ typedef struct ValkeyModuleAtomicSlotMigrationInfo {
+    int is_module_client;      /* 1 if command was from RM_Call, 0 otherwise. */
+    int argc;                  /* Number of command arguments. */
+    ValkeyModuleString **argv; /* Command arguments array (zero-copy). */
+    const char *object;        /* Rejection context: ACL_KEY/ACL_CHANNEL=denied key or channel,
```
If only used for rejection, should we change the name to reflect that?

```suggestion
    const char *rejection_context;        /* Rejection context: ACL_KEY/ACL_CHANNEL=denied key or channel,
```

martinrvisser replied:
Done

murphyjacob4 commented on src/server.c:
```diff
@@ -4315,6 +4343,14 @@ int processCommand(client *c) {
             clusterRedirectClient(c, n, c->slot, error_code);
             c->duration = 0;
             c->cmd->rejected_calls++;
+            char cluster_target[NET_IP_STR_LEN + 16] = {0};
```
Maybe we can just supply the slot as context. The module should be able to figure out what node that maps to if it wants.

martinrvisser replied:
Good idea, done!

madolson commented on src/valkeymodule.h:
```diff
@@ -674,6 +680,27 @@ static const ValkeyModuleEvent ValkeyModuleEvent_ReplicationRoleChanged = {VALKE
+#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_NOAUTH 0
+#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_ACL_CMD 1
+#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_ACL_KEY 2
+#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_ACL_CHANNEL 3
```
We also need to support the per database rejection.

madolson commented on src/valkeymodule.h:
```diff
@@ -674,6 +680,27 @@ static const ValkeyModuleEvent ValkeyModuleEvent_ReplicationRoleChanged = {VALKE
 #define VALKEYMODULE_SUBEVENT_ATOMIC_SLOT_MIGRATION_EXPORT_COMPLETED 5
 #define _VALKEYMODULE_SUBEVENT_ATOMIC_SLOT_MIGRATION_NEXT 6
 
+#define VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_NOAUTH 0
```
During the weekly meeting we discussed if it was possible to:
1. Merge All the ACL errors into the same enums as the existing codes exposed by ACL log.
2. If we needed all of the other command error codes. There are a lot to keep track of and we will need to continue adding more in the future. Maybe just an "Other" category would be sufficient?

martinrvisser replied:
@madolson 
Re 1: makes sense to reuse the ACL_LOG enums. 
Re 2: can we clarify this, is the objective to reduce the number down or just add an OTHER category?

madolson reviewed (commented):
@martinrvisser We discussed this a bit in the weekly meeting. We wanted to get a deeper understanding of what features that auditing module needed to make sure we're building the right set of APIs for it. We feel it's a bit late to add it for 9.1. Do you have some implementation that would build against this?

martinrvisser committed:
adjusted comments, removed PR number, renamed object to rejection_context and changed to cluster slot from host:port

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
consolidated non-ACL subevents into OTHER with rejection_context providing the detail

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
Fix stale error code examples and NOMULTI test using correct CMD_NO_MULTI command

- Replace "REDIRECT"/"ERR" examples with "CLUSTERDOWN"/"NOMULTI" in docblock
- Update moduleFireCommandRejectedEvent comment to use "NOMULTI" example
- Fix NOMULTI test to use MULTI-inside-MULTI (has CMD_NO_MULTI flag) instead
  of SUBSCRIBE which gets queued rather than rejected
- Fix command_name -> command dict key in NOMULTI/PUBSUB/NOREPLICAS tests

Signed-off-by: martinrvisser <mvisser@hotmail.com>

murphyjacob4 commented on src/server.c:
```diff
@@ -4389,6 +4411,7 @@ int processCommand(client *c) {
 
     if (c->flag.multi && c->cmd->flags & CMD_NO_MULTI) {
         rejectCommandFormat(c, "Command '%s' not allowed inside a transaction", c->cmd->fullname);
+        moduleFireCommandRejectedEvent(c, VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_OTHER, -1, "NOMULTI");
```
Would it be better to just take the whole reply we are sending the user and just pass that to the module? That way we don't special codes ("NOMULTI") - the module could just parse the rejection string and categorize it.

martinrvisser replied:
That makes perfect sense, thanks

murphyjacob4 commented on src/module.c:
```diff
@@ -12466,6 +12588,70 @@ static uint64_t moduleEventVersions[] = {
+ *       - VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_OTHER:
+ *         the Valkey error code string (e.g. "OOM", "LOADING", "BUSY",
+ *         "MOVED", "ASK", "CLUSTERDOWN", "NOREPLICAS", "NOMULTI", ...)
+ *       - All other ACL subevents: NULL
```
Seeing it written out, it seems like we really have two unique events here:

1. ACL rejection event - subevent is ValkeyModuleACLLogEntryReason, we supply the denied key/chanel in the event
2. Other rejection event - no subevent breakdown, we supply the actual error reply and make the module parse it

Given the struct contents of the two events are fairly distinct, would it be cleaner to make them two different high level events?

On the same note, since `VALKEYMODULE_SUBEVENT_COMMAND_RESULT_REJECTED_OTHER = 5`, I don't think we can cleanly add another ACL log type after `VALKEYMODULE_ACL_LOG_DB = 4` cleanly. Using two different events would help separate those.

martinrvisser replied:
Agreed that would cleaner

martinrvisser replied:
@murphyjacob4 thoughts? : 
VALKEYMODULE_EVENT_COMMAND_RESULT_REJECTED **or** VALKEYMODULE_EVENT_COMMAND_RESULT_REJECTED_OTHER **or** VALKEYMODULE_EVENT_COMMAND_RESULT_NONACL_REJECTED 
```
#define VALKEYMODULE_EVENT_COMMAND_RESULT_SUCCESS 20
#define VALKEYMODULE_EVENT_COMMAND_RESULT_FAILURE 21
#define VALKEYMODULE_EVENT_COMMAND_RESULT_REJECTED 22
#define VALKEYMODULE_EVENT_COMMAND_RESULT_ACL_REJECTED 23
```

murphyjacob4 replied:
Good question. Naming is hard

I would prefer

```
#define VALKEYMODULE_EVENT_COMMAND_RESULT_SUCCESS      20
#define VALKEYMODULE_EVENT_COMMAND_RESULT_FAILURE      21
#define VALKEYMODULE_EVENT_COMMAND_RESULT_ACL_REJECTED 22
#define VALKEYMODULE_EVENT_COMMAND_RESULT_REJECTED     23
```

The documentation will have to do some lifting, but I'm okay with that.

@martinrvisser

murphyjacob4 reviewed (commented):
Thank you for your patience! This module API is one of the harder ones to design, so I have some additional points of discussion.

martinrvisser committed:
two high level events ACL rejected and non-ACL rejected, expose full error message for non-ACL

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
fix for cluster_err_msg

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
fix double string termination

Signed-off-by: martinrvisser <mvisser@hotmail.com>

murphyjacob4 commented on src/server.c:
```diff
@@ -4558,10 +4593,12 @@ int processCommand(client *c) {
-                            "Can't execute '%s': only (P|S)SUBSCRIBE / "
-                            "(P|S)UNSUBSCRIBE / PING / QUIT / RESET are allowed in this context",
-                            c->cmd->fullname);
+        sds pubsub_err = sdscatprintf(sdsempty(),
+                                      "Can't execute '%s': only (P|S)SUBSCRIBE / "
+                                      "(P|S)UNSUBSCRIBE / PING / QUIT / RESET are allowed in this context",
+                                      c->cmd->fullname);
+        sdsmapchars(pubsub_err, "\r\n", "  ", 2);
```
nit: Do you think it would be cleaner to just add the moduleFireCommandRejectedEvent to rejectCommand and rejectCommandSds (which is also used by rejectCommandFormat) - that way the call sites will remain unchanged? Also would mean we can keep the sdsmapchars inside the helper for rejectCommandFormat

So effectively make rejectCommandSds the same as what processCommandRejectSds does now

I guess AUTH uses rejectCommand as well, so maybe we just add a flag to rejectCommand for notify_modules, so the AUTH paths can specify false and do their own notification?

murphyjacob4 commented on src/module.c:
```diff
@@ -390,6 +390,7 @@ typedef struct ValkeyModuleCommandFilter {
 /* Registered filters */
 static list *moduleCommandFilters;
 
+
```
nit: unneeded extra newline

murphyjacob4 reviewed (approved):
I think we landed in a good place. Thanks @martinrvisser for your patience

I think core team approved it per this comment:

> Discussed in the meeting to simplify and just return the error codes. If it's updated this week we will merge it in for RC2.

But will mention the team here so we are all aware @valkey-io/core-team

martinrvisser committed:
removed wrapper functions in processCommand and moved into rejectCommand and rejectCommandSds

Signed-off-by: martinrvisser <mvisser@hotmail.com>

sarthakaggarwal97 commented:
I am seeing test failures in PR CIs post rebasing with unstable:
1. https://github.com/valkey-io/valkey/actions/runs/24735227490/job/72360139943?pr=3504#step:9:1306
2. https://github.com/valkey-io/valkey/actions/runs/24735227490/job/72360139953?pr=3504#step:10:852
